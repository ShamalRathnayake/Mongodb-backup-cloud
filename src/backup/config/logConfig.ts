export default {
  types: {
    error: 'error',
    warn: 'warn',
    info: 'info',
    debug: 'debug',
  },
  backupLogPath: 'mongodb-backup-cloud-logs/backup',
  uploadLogPath: 'mongodb-backup-cloud-logs/upload',
};
