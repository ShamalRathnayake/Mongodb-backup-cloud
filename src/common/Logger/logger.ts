/* eslint-disable max-classes-per-file */
import { createLogger, format, Logger } from 'winston';
import DailyRotateFile from 'winston-daily-rotate-file';

import logConfig from '../../backup/config/logConfig';

abstract class BaseLogger {
  protected logger: Logger | null;

  protected isLoggingEnabled: boolean;

  protected constructor() {
    this.logger = null;
    this.isLoggingEnabled = true;
  }

  public toggleLogging(status: boolean): void {
    this.isLoggingEnabled = status;
    if (!status) {
      this.logger = null;
      return;
    }

    const transportOptions = {
      filename: this.getLogFileName(),
      dirname: this.getLogFilePath(),
      datePattern: 'YYYY-MM-DD',
      zippedArchive: true,
      maxSize: '20m',
      maxFiles: '14d',
      handleExceptions: true,
      exitOnError: false,
    };

    this.logger = createLogger({
      level: 'debug',
      transports: [new DailyRotateFile(transportOptions)],
      format: format.combine(
        format.timestamp({
          format: 'YYYY-MM-DD HH:mm:ss A',
        }),
        format.printf(({ timestamp, level, message }) => {
          return `${'-'.repeat(
            100
          )}\n[${timestamp}] - [${this.getServiceName()}] - ${level}: ${message}`;
        }),
        format.colorize(),
        format.align()
      ),
    });
  }

  protected abstract getLogFileName(): string;

  protected abstract getLogFilePath(): string;

  protected abstract getServiceName(): string;

  public log(level: string, message: string): void {
    if (this.isLoggingEnabled && this.logger) {
      this.logger.log(level, message);
    }
  }
}

export class BackupLogger extends BaseLogger {
  // eslint-disable-next-line no-use-before-define
  private static instance: BackupLogger = new BackupLogger();

  private constructor() {
    super();
  }

  protected getLogFileName(): string {
    // eslint-disable-next-line quotes
    return `backup-%DATE%.log`;
  }

  protected getLogFilePath(): string {
    return logConfig.backupLogPath;
  }

  protected getServiceName(): string {
    return 'backup-service';
  }

  public static toggleLogging(status: boolean): void {
    BackupLogger.instance.toggleLogging(status);
  }

  public static log(level: string, message: string): void {
    BackupLogger.instance.log(level, message);
  }
}

export class UploadLogger extends BaseLogger {
  // eslint-disable-next-line no-use-before-define
  private static instance: UploadLogger = new UploadLogger();

  private constructor() {
    super();
  }

  protected getLogFileName(): string {
    // eslint-disable-next-line quotes
    return `upload-%DATE%.log`;
  }

  protected getLogFilePath(): string {
    return logConfig.uploadLogPath;
  }

  protected getServiceName(): string {
    return 'upload-service';
  }

  public static toggleLogging(status: boolean): void {
    UploadLogger.instance.toggleLogging(status);
  }

  public static log(level: string, message: string): void {
    UploadLogger.instance.log(level, message);
  }
}
