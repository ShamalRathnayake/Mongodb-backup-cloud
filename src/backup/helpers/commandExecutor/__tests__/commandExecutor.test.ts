/* eslint-disable */
// @ts-nocheck

import childProcess, { ExecException, exec } from 'child_process';
import CommandExecutor from '../commandExecutor';
import { BackupLogger } from '../../../../common/Logger/logger';
import logConfig from '../../../config/logConfig';

jest.mock('child_process');
const mockedExec = exec as jest.MockedFunction<typeof exec>;

jest.mock('../../../../common/Logger/logger');

describe('Command Executor', () => {
  afterEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
  });

  it('executes the command and returns correct status and log on success', async () => {
    // Test implementation
    const command = 'echo Hello World!';
    const stdout = 'Hello World!';
    const stderr = '';

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(true);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(stdout);
  });

  it('handles command execution failure and returns correct status and log', async () => {
    // Test implementation
    const command = 'echo Hello World!';
    const stdout = 'Hello World!';
    const stderr = '';

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(true);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(stdout);
  });

  it('handles non-Error type error and returns correct status and log', async () => {
    // Test implementation
    const command = 'some-command';
    const nonErrorString = 'Non-Error error message';
    const stderr = 'Error output message';

    // Mock the exec function to simulate non-Error type error
    mockedExec.mockImplementation((_command, callback) => {
      callback(nonErrorString, { stdout: '', stderr });
    });

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(false);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(nonErrorString);
  });

  it('logs debug messages on successful command execution', async () => {
    // Test implementation
    const command = 'echo Hello World!';
    const stdout = 'Hello World!';
    const stderr = '';

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(true);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(stdout);
    expect(BackupLogger.log).toHaveBeenCalledWith(
      logConfig.types.debug,
      'Command execution started'
    );
    expect(BackupLogger.log).toHaveBeenCalledWith(
      logConfig.types.debug,
      'Command execution finished'
    );
  });

  it('logs error messages on command execution failure', async () => {
    // Test implementation
    const command = 'some-command';
    const stderr = 'Error output message';

    // Mock the exec function to simulate non-Error type error
    mockedExec.mockImplementation((_command, callback) => {
      callback(Error(stderr), { stdout: '', stderr });
    });

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(false);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(stderr);
    expect(BackupLogger.log).toHaveBeenCalledWith(
      logConfig.types.debug,
      'Command execution started'
    );
    expect(BackupLogger.log).toHaveBeenCalledWith(logConfig.types.error, stderr);
  });

  // Additional integration testing scenarios

  it('correctly logs debug messages when logger is mocked', async () => {
    // Test implementation
    const command = 'echo Hello World!';
    const stdout = 'Hello World!';
    const stderr = '';

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });

    // Mock the BackupLogger.log function
    const mockedLog = BackupLogger.log.mockImplementation(() => {});

    const result = await CommandExecutor.executeCommand(command);

    expect(mockedExec).toHaveBeenCalledWith(command, expect.any(Function));
    expect(mockedExec).toBeCalledTimes(1);
    expect(result).toHaveProperty('status');
    expect(result.status).toBe(true);
    expect(result).toHaveProperty('log');
    expect(result.log).toBe(stdout);
    expect(mockedLog).toHaveBeenCalledWith(logConfig.types.debug, 'Command execution started');
    expect(mockedLog).toHaveBeenCalledWith(logConfig.types.debug, 'Command execution finished');
  });

  // Snapshot testing scenarios

  it('matches snapshot for logConfig', async () => {
    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });
    // Test implementation
    expect(logConfig).toMatchSnapshot();
  });

  it('matches snapshot for executeCommand output', async () => {
    // Test implementation
    const command = 'echo Hello World!';

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout, stderr });
    });

    const result = await CommandExecutor.executeCommand(command);
    expect(result).toMatchSnapshot();
  });

  // Additional edge testing scenarios

  it('correctly handles parallel command execution', async () => {
    // Test implementation
    // Mock the BackupLogger.log function
    const mockedLog = BackupLogger.log.mockImplementation(() => {});

    // Replace with the actual test commands
    const commands = ['echo Command_1', 'echo Command_2', 'echo Command_3'];

    // Mock the exec function to simulate successful execution
    mockedExec.mockImplementation((_command, callback) => {
      callback(null, { stdout: '', stderr: '' });
    });

    const executePromises = commands.map((command) => CommandExecutor.executeCommand(command));

    // Simulate parallel execution
    await Promise.all(executePromises);

    // Verify that the Logger was called with the correct logs
    expect(mockedLog).toBeCalledTimes(6);
    expect(mockedLog).nthCalledWith(1, expect.anything(), 'Command execution started');
    expect(mockedLog).nthCalledWith(4, expect.anything(), 'Command execution finished');

    // Restore the original Logger implementation
    jest.restoreAllMocks();
  });
});
