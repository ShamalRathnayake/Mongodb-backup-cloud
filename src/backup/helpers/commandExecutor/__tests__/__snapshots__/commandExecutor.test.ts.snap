// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Command Executor matches snapshot for executeCommand output 1`] = `
{
  "log": "stdout is not defined",
  "status": false,
}
`;

exports[`Command Executor matches snapshot for logConfig 1`] = `
{
  "backupLogPath": "mongodb-backup-cloud-logs/backup",
  "types": {
    "debug": "debug",
    "error": "error",
    "info": "info",
    "warn": "warn",
  },
  "uploadLogPath": "mongodb-backup-cloud-logs/upload",
}
`;
